diff --git a/arch/x86/entry/syscalls/syscall_64.tbl b/arch/x86/entry/syscalls/syscall_64.tbl
index c84d12608cd2..902d21de4cae 100644
--- a/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/arch/x86/entry/syscalls/syscall_64.tbl
@@ -343,6 +343,7 @@
 332	common	statx			sys_statx
 333	common	io_pgetevents		sys_io_pgetevents
 334	common	rseq			sys_rseq
+335 common syscall         sys_hypercall
 # don't use numbers 387 through 423, add new calls after the last
 # 'common' entry
 424	common	pidfd_send_signal	sys_pidfd_send_signal
@@ -417,3 +418,8 @@
 547	x32	pwritev2		compat_sys_pwritev64v2
 # This is the end of the legacy x32 range.  Numbers 548 and above are
 # not special and are not to be used for x32-specific syscalls.
+
+548>---64>-----mycall>->------->-------sys_mycall
+549>---64>-----mycall2>>------->-------sys_mycall2
+550>---64>-----mycall3>>------->-------sys_mycall3
+
diff --git a/arch/x86/kernel/Makefile b/arch/x86/kernel/Makefile
index 96d51bbc2bd4..83f37f8eb0ae 100644
--- a/arch/x86/kernel/Makefile
+++ b/arch/x86/kernel/Makefile
@@ -25,6 +25,10 @@ KASAN_SANITIZE_dumpstack_$(BITS).o			:= n
 KASAN_SANITIZE_stacktrace.o				:= n
 KASAN_SANITIZE_paravirt.o				:= n
 KASAN_SANITIZE_sev.o					:= n
+# Prevents double fault generated due to inducing
+# # a second page fault while serving the first
+KASAN_SANITIZE_traps.o					:= n
+
 
 # With some compiler versions the generated code results in boot hangs, caused
 # by several compilation units. To be safe, disable all instrumentation.
diff --git a/arch/x86/kernel/cpu/Makefile b/arch/x86/kernel/cpu/Makefile
index d7e3ceaf75c1..96701c1e6f43 100644
--- a/arch/x86/kernel/cpu/Makefile
+++ b/arch/x86/kernel/cpu/Makefile
@@ -13,6 +13,25 @@ endif
 KCOV_INSTRUMENT_common.o := n
 KCOV_INSTRUMENT_perf_event.o := n
 KMSAN_SANITIZE_common.o := n
+# If this file is instrumented, it causes a boot-loop.
+# Normally, KASAN sanitizing this file wouldn't be an issue.
+# However, we hijack KASAN hooks to inject EVTRACK instrumentation.
+# When the crash takes place, the call stack looks like this:
+#
+#   0xffffffff84391d37 <setup_per_cpu_areas+754>    call   switch_to_new_gdt
+#   0xffffffff81058734 <switch_to_new_gdt+132>      call   load_percpu_segment
+#   0xffffffff81058678 <load_percpu_segment+24>     call   __asan_load8_noabort·
+#   0xffffffff813d3c63 <__asan_load8_noabort+51>    call   record_event
+#
+#   0xffffffff8124d400 <record_event>        push   r15
+#   0xffffffff8124d402 <record_event+2>      push   r14
+# ► 0xffffffff8124d404 <record_event+4>      mov    eax, dword ptr gs:[rip + 0x7edd3afd]		# 0x20f08 <__preempt_count>
+# The memory access at the `mov` instruction causes a panic.
+# In Linux x86-64, `gs` register is used to access per-cpu variables.
+# The panic occurrs when the EVTRACK trampoline attemps to access `__preempt_count`.
+# We exclude `common.o` from KASAN sanitization just to avoid
+# instrumenting the `load_percpu_segment()` method.
+KASAN_SANITIZE_common.o := n
 
 # As above, instrumenting secondary CPU boot code causes boot hangs.
 KCSAN_SANITIZE_common.o := n
diff --git a/arch/x86/mm/Makefile b/arch/x86/mm/Makefile
index c80febc44cd2..a67f49e75af4 100644
--- a/arch/x86/mm/Makefile
+++ b/arch/x86/mm/Makefile
@@ -10,6 +10,10 @@ KASAN_SANITIZE_mem_encrypt.o		:= n
 KASAN_SANITIZE_mem_encrypt_amd.o	:= n
 KASAN_SANITIZE_mem_encrypt_identity.o	:= n
 KASAN_SANITIZE_pgprot.o		:= n
+# Prevents double fault generated due to inducing
+# a second page fault while serving the first
+KASAN_SANITIZE_fault.o		:= n
+
 
 # Disable KCSAN entirely, because otherwise we get warnings that some functions
 # reference __initdata sections.
diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index 3aa04224315e..5f25342c6348 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -19,6 +19,7 @@
 #include <linux/sched/mm.h>
 #include <asm/unaligned.h>
 #include <crypto/hash.h>
+#include <linux/delay.h>
 #include "ctree.h"
 #include "disk-io.h"
 #include "transaction.h"
@@ -4680,6 +4681,7 @@ void __cold close_ctree(struct btrfs_fs_info *fs_info)
 	 * still try to wake up the cleaner.
 	 */
 	kthread_park(fs_info->cleaner_kthread);
+	printk(KERN_INFO "[PID: %d] close_ctree: kthread_park\n", current->pid);
 
 	/* wait for the qgroup rescan worker to stop */
 	btrfs_qgroup_wait_for_completion(fs_info, false);
@@ -4765,6 +4767,7 @@ void __cold close_ctree(struct btrfs_fs_info *fs_info)
 
 	kthread_stop(fs_info->transaction_kthread);
 	kthread_stop(fs_info->cleaner_kthread);
+	printk(KERN_INFO "[PID: %d] close_ctree: kthread_stop\n", current->pid);
 
 	ASSERT(list_empty(&fs_info->delayed_iputs));
 	set_bit(BTRFS_FS_CLOSING_DONE, &fs_info->flags);
@@ -4797,6 +4800,7 @@ void __cold close_ctree(struct btrfs_fs_info *fs_info)
 	 */
 	invalidate_inode_pages2(fs_info->btree_inode->i_mapping);
 	btrfs_stop_all_workers(fs_info);
+	printk(KERN_INFO "[PID: %d] close_ctree: btrfs_stop_all_workers\n", current->pid);
 
 	/* We shouldn't have any transaction open at this point */
 	warn_about_uncommitted_trans(fs_info);
diff --git a/fs/btrfs/extent_io.c b/fs/btrfs/extent_io.c
index 3bbf8703db2a..2b49c98beb00 100644
--- a/fs/btrfs/extent_io.c
+++ b/fs/btrfs/extent_io.c
@@ -1985,6 +1985,8 @@ static noinline_for_stack int __extent_writepage_io(struct btrfs_inode *inode,
 	bool has_error = false;
 	bool compressed;
 
+	printk(KERN_INFO "[PID: %d] __extent_writepage_io: begin\n", current->pid);
+
 	ret = btrfs_writepage_cow_fixup(page);
 	if (ret) {
 		/* Fixup worker will requeue */
@@ -2146,6 +2148,8 @@ static int __extent_writepage(struct page *page, struct writeback_control *wbc,
 	loff_t i_size = i_size_read(inode);
 	unsigned long end_index = i_size >> PAGE_SHIFT;
 
+	printk(KERN_INFO "[PID: %d] __extent_writepage: begin\n", current->pid);
+
 	trace___extent_writepage(page, inode, wbc);
 
 	WARN_ON(!PageLocked(page));
@@ -3010,6 +3014,7 @@ static int extent_write_cache_pages(struct address_space *mapping,
 	 * anyway and we'd rather not waste our time writing out stuff that is
 	 * going to be truncated anyway.
 	 */
+	printk(KERN_INFO "[PID: %d] extent_write_cache_pages: begin\n", current->pid);
 	if (!igrab(inode))
 		return 0;
 
@@ -3124,7 +3129,8 @@ static int extent_write_cache_pages(struct address_space *mapping,
 
 	if (wbc->range_cyclic || (wbc->nr_to_write > 0 && range_whole))
 		mapping->writeback_index = done_index;
-
+	
+	printk(KERN_INFO "[PID: %d] extent_write_cache_pages: btrfs_add_delayed_iput\n", current->pid);
 	btrfs_add_delayed_iput(BTRFS_I(inode));
 	return ret;
 }
diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index 98a800b8bd43..c161ab0e33cf 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -34,6 +34,7 @@
 #include <linux/iomap.h>
 #include <asm/unaligned.h>
 #include <linux/fsverity.h>
+#include <linux/delay.h>
 #include "misc.h"
 #include "ctree.h"
 #include "disk-io.h"
@@ -1493,6 +1494,7 @@ static noinline void async_cow_start(struct btrfs_work *work)
 
 	compressed_extents = compress_file_range(async_chunk);
 	if (compressed_extents == 0) {
+		printk(KERN_INFO "[PID: %d] async_cow_start: btrfs_add_delayed_iput\n", current->pid);
 		btrfs_add_delayed_iput(async_chunk->inode);
 		async_chunk->inode = NULL;
 	}
@@ -2903,6 +2905,8 @@ static void btrfs_writepage_fixup_worker(struct btrfs_work *work)
 	int ret = 0;
 	bool free_delalloc_space = true;
 
+	printk(KERN_INFO "[PID: %d] btrfs_writepage_fixup_worker: begin\n", current->pid);
+
 	fixup = container_of(work, struct btrfs_writepage_fixup, work);
 	page = fixup->page;
 	inode = fixup->inode;
@@ -3015,6 +3019,7 @@ static void btrfs_writepage_fixup_worker(struct btrfs_work *work)
 	 * that could need flushing space. Recursing back to fixup worker would
 	 * deadlock.
 	 */
+	printk(KERN_INFO "[PID: %d] btrfs_writepage_fixup_worker: btrfs_add_delayed_iput\n", current->pid);
 	btrfs_add_delayed_iput(inode);
 }
 
@@ -3035,9 +3040,14 @@ int btrfs_writepage_cow_fixup(struct page *page)
 	struct btrfs_fs_info *fs_info = btrfs_sb(inode->i_sb);
 	struct btrfs_writepage_fixup *fixup;
 
+	printk(KERN_INFO "[PID: %d] btrfs_writepage_cow_fixup: begin\n", current->pid);
+
 	/* This page has ordered extent covering it already */
-	if (PageOrdered(page))
+	if (PageOrdered(page)){
+		printk(KERN_INFO "[PID: %d] btrfs_writepage_cow_fixup: PageOrdered(page): %d\n", current->pid, PageOrdered(page));
 		return 0;
+	}
+		
 
 	/*
 	 * PageChecked is set below when we create a fixup worker for this page,
@@ -3592,6 +3602,7 @@ unsigned int btrfs_verify_data_csum(struct btrfs_bio *bbio,
 void btrfs_add_delayed_iput(struct btrfs_inode *inode)
 {
 	struct btrfs_fs_info *fs_info = inode->root->fs_info;
+	printk(KERN_INFO "[PID: %d] btrfs_add_delayed_iput: begin\n", current->pid);
 
 	if (atomic_add_unless(&inode->vfs_inode.i_count, -1, 1))
 		return;
@@ -3601,8 +3612,11 @@ void btrfs_add_delayed_iput(struct btrfs_inode *inode)
 	ASSERT(list_empty(&inode->delayed_iput));
 	list_add_tail(&inode->delayed_iput, &fs_info->delayed_iputs);
 	spin_unlock(&fs_info->delayed_iput_lock);
-	if (!test_bit(BTRFS_FS_CLEANER_RUNNING, &fs_info->flags))
+	if (!test_bit(BTRFS_FS_CLEANER_RUNNING, &fs_info->flags)){
+		printk(KERN_INFO "[PID: %d] btrfs_add_delayed_iput: wake_up_process\n", current->pid);
 		wake_up_process(fs_info->cleaner_kthread);
+	}
+		
 }
 
 static void run_delayed_iput_locked(struct btrfs_fs_info *fs_info,
diff --git a/fs/f2fs/file.c b/fs/f2fs/file.c
index ecbc8c135b49..59859cd76989 100644
--- a/fs/f2fs/file.c
+++ b/fs/f2fs/file.c
@@ -2254,6 +2254,7 @@ static int f2fs_ioc_shutdown(struct file *filp, unsigned long arg)
 		goto out;
 	}
 
+	printk(KERN_INFO "[PID: %d] f2fs_ioc_shutdown: f2fs_stop_gc_thread.\n", current->pid);
 	f2fs_stop_gc_thread(sbi);
 	f2fs_stop_discard_thread(sbi);
 
diff --git a/fs/f2fs/gc.c b/fs/f2fs/gc.c
index 6e2cae3d2e71..b64faf2854cb 100644
--- a/fs/f2fs/gc.c
+++ b/fs/f2fs/gc.c
@@ -206,12 +206,17 @@ int f2fs_start_gc_thread(struct f2fs_sb_info *sbi)
 void f2fs_stop_gc_thread(struct f2fs_sb_info *sbi)
 {
 	struct f2fs_gc_kthread *gc_th = sbi->gc_thread;
+	printk(KERN_INFO "[PID: %d] f2fs_stop_gc_thread: begin.\n",current->pid);
 
-	if (!gc_th)
+	if (!gc_th){
+		printk(KERN_INFO "[PID: %d] f2fs_stop_gc_thread: gc_th is null.\n",current->pid);
 		return;
+	}
+		
 	kthread_stop(gc_th->f2fs_gc_task);
 	wake_up_all(&gc_th->fggc_wq);
 	kfree(gc_th);
+	printk(KERN_INFO "[PID: %d] f2fs_stop_gc_thread: free gc_th.\n",current->pid);
 	sbi->gc_thread = NULL;
 }
 
diff --git a/fs/f2fs/super.c b/fs/f2fs/super.c
index 1f812b9ce985..f7783c206d64 100644
--- a/fs/f2fs/super.c
+++ b/fs/f2fs/super.c
@@ -670,6 +670,8 @@ static int parse_options(struct super_block *sb, char *options, bool is_remount)
 	kgid_t gid;
 	int ret;
 
+	printk(KERN_INFO "[PID: %d] parse_options: opt: %s\n",current->pid, options);
+
 	if (!options)
 		goto default_check;
 
@@ -2359,6 +2361,7 @@ static int f2fs_remount(struct super_block *sb, int *flags, char *data)
 			(F2FS_OPTION(sbi).bggc_mode == BGGC_MODE_OFF &&
 			!test_opt(sbi, GC_MERGE))) {
 		if (sbi->gc_thread) {
+			printk(KERN_INFO "[PID: %d] f2fs_remount:f2fs_stop_gc_thread 1.\n",current->pid);
 			f2fs_stop_gc_thread(sbi);
 			need_restart_gc = true;
 		}
@@ -2475,6 +2478,7 @@ static int f2fs_remount(struct super_block *sb, int *flags, char *data)
 		if (f2fs_start_gc_thread(sbi))
 			f2fs_warn(sbi, "background gc thread has stopped");
 	} else if (need_stop_gc) {
+		printk(KERN_INFO "[PID: %d] f2fs_remount:f2fs_stop_gc_thread 2.\n",current->pid);
 		f2fs_stop_gc_thread(sbi);
 	}
 restore_opts:
@@ -4488,6 +4492,7 @@ static int f2fs_fill_super(struct super_block *sb, void *data, int silent)
 	if ((F2FS_OPTION(sbi).bggc_mode != BGGC_MODE_OFF ||
 		test_opt(sbi, GC_MERGE)) && !f2fs_readonly(sb)) {
 		/* After POR, we can run background GC thread.*/
+		printk(KERN_INFO "[PID: %d] f2fs_fill_super: f2fs_start_gc_thread.\n", current->pid);
 		err = f2fs_start_gc_thread(sbi);
 		if (err)
 			goto sync_free_meta;
diff --git a/include/linux/evtrack.h b/include/linux/evtrack.h
new file mode 100644
index 000000000000..e900e3d75abd
--- /dev/null
+++ b/include/linux/evtrack.h
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _LINUX_EVTRACK_H
+#define _LINUX_EVTRACK_H
+
+#include <uapi/linux/evtrack.h>
+
+struct task_struct;
+
+#ifdef CONFIG_EVTRACK
+
+enum evtrack_mode {
+	/* Coverage collection is not enabled yet. */
+	EVTRACK_MODE_DISABLED = 0,
+	/* EVTRACK was initialized, but tracing mode hasn't been chosen yet. */
+	EVTRACK_MODE_INIT = 1,
+	/* Track all the allocations */
+	EVTRACK_MODE_TRACK_ALL = 2,
+	/* Temporarily stop event tracing */
+	EVTRACK_MODE_STOP = 3
+};
+
+/*
+ * evtrack descriptor (one per opened debugfs file).
+ * State transitions of the descriptor:
+ *  - initial state after open()
+ *  - then there must be a single ioctl(EVTRACK_INIT_TRACE) call
+ *  - then, mmap() call (several calls are allowed but not useful)
+ *  - then, ioctl(EVTRACK_ENABLE, arg), where arg is
+ *  - then, ioctl(EVTRACK_DISABLE) to disable the task.
+ * Enabling/disabling ioctls can be repeated (only one task a time allowed).
+ */
+struct evtrack {
+    /*
+     * Reference counter. We keep one for:
+     *  - opened file descriptor
+     *  - task with enabled event tracking (we can't unwire it from another task)
+     */
+    refcount_t		refcount;
+    /* The lock protects mode, nr_events, area and t. */
+    spinlock_t		lock;
+    enum evtrack_mode		mode;
+    /* Number of events to be recorded. */
+    unsigned int		nr_events;
+    /* Coverage buffer shared with user space. */
+    void			*area;
+    /* Task for which we collect event, or NULL. */
+    struct task_struct	*t;
+    /* RB-tree root for storing seen allocations. */
+    struct rb_root root;
+    /* Timestamp of first event recorded during this system call */
+    __u64 timestamp;
+};
+
+struct api {
+    char name[256];
+    unsigned long start, size;
+};
+
+void evtrack_task_init(struct task_struct *t);
+void evtrack_task_exit(struct task_struct *t);
+void record_event(enum evtrack_event_types event_type, const void* ptr, size_t size, bool keep_tag);
+
+#else
+
+static inline void evtrack_task_init(struct task_struct *t) {}
+static inline void evtrack_task_exit(struct task_struct *t) {}
+static void record_event(enum evtrack_event_types event_type, const void* ptr, size_t size, bool keep_tag) {}
+
+#endif /* CONFIG_EVTRACK */
+#endif /* _LINUX_EVTRACK_H */
diff --git a/include/linux/mycall.h b/include/linux/mycall.h
new file mode 100644
index 000000000000..3fb4cf1ddf8c
--- /dev/null
+++ b/include/linux/mycall.h
@@ -0,0 +1,3 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+char *get_my_pointer(void);
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 853d08f7562b..78f6cc7e2585 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1423,6 +1423,21 @@ struct task_struct {
 	unsigned int			kcov_softirq;
 #endif
 
+#ifdef CONFIG_EVTRACK
+	/* Event tracking enabled for this task (0 if disabled): */
+	unsigned int			evtrack_mode;
+
+	/* Size of the evtrack_area: */
+	unsigned int			evtrack_size;
+
+	/* Buffer for event tracking: */
+	void				*evtrack_area;
+
+	/* EVTRACK descriptor wired with this task or NULL: */
+	struct evtrack			*evtrack;
+#endif
+
+
 #ifdef CONFIG_MEMCG
 	struct mem_cgroup		*memcg_in_oom;
 	gfp_t				memcg_oom_gfp_mask;
diff --git a/include/linux/syscalls.h b/include/linux/syscalls.h
index 33a0ee3bcb2e..74e0dbc5a2e6 100644
--- a/include/linux/syscalls.h
+++ b/include/linux/syscalls.h
@@ -1361,6 +1361,10 @@ static inline unsigned int ksys_personality(unsigned int personality)
 	return old;
 }
 
+asmlinkage long sys_mycall(void);
+asmlinkage long sys_mycall2(void);
+asmlinkage long sys_mycall3(void);
+
 /* for __ARCH_WANT_SYS_IPC */
 long ksys_semtimedop(int semid, struct sembuf __user *tsops,
 		     unsigned int nsops,
@@ -1388,3 +1392,4 @@ int __sys_getsockopt(int fd, int level, int optname, char __user *optval,
 int __sys_setsockopt(int fd, int level, int optname, char __user *optval,
 		int optlen);
 #endif
+asmlinkage long sys_hypercall(unsigned long id, unsigned long cmd, unsigned long prog_num, unsigned long sched);
\ No newline at end of file
diff --git a/include/uapi/linux/evtrack.h b/include/uapi/linux/evtrack.h
new file mode 100644
index 000000000000..3b8d46b84106
--- /dev/null
+++ b/include/uapi/linux/evtrack.h
@@ -0,0 +1,67 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
+#ifndef _LINUX_EVTRACK_IOCTLS_H
+#define _LINUX_EVTRACK_IOCTLS_H
+
+#include <linux/types.h>
+#include <linux/rbtree.h>
+#include <linux/atomic.h>
+
+#define EVTRACK_INIT_TRACK                     _IOR('c', 1, unsigned long)
+#define EVTRACK_ENABLE                     _IO('c', 100)
+#define EVTRACK_DISABLE                            _IO('c', 101)
+#define EVTRACK_DEBUG                      _IO('c', 102)
+#define EVTRACK_STOP                       _IO('c', 103)
+
+/* Maximum number of stack trace entries in the context stored */
+#define NR_MAX_TRACE_ENTRIES        32
+
+enum {
+       /* Track all the allocations */
+       EVTRACK_TRACK_ALL = 0
+};
+
+enum evtrack_event_types {
+       /* Heap allocations, e.g. kmalloc() */
+       EVTRACK_EVENT_HEAP_ALLOCATION = 1,
+
+       /* Heap deallocations, e.g. kfree() */
+       EVTRACK_EVENT_HEAP_DEALLOCATION = 2,
+
+       /* Reading from heap buffer */
+       EVTRACK_EVENT_HEAP_READ = 3,
+
+       /* Writing to heap buffer */
+       EVTRACK_EVENT_HEAP_WRITE = 4
+};
+
+struct evtrack_event {
+       unsigned int event_id;                    // Monotonically increasing event id
+       enum evtrack_event_types type;    // Type of the event
+       const void* ptr;                  // Pointer associated to the event
+	   const void* alloc_ptr;
+       size_t size;                      // Size associated to the event
+       unsigned int nr_trace;                    // Number of entries in the stack trace
+       /*
+        * Ref: http://static.lwn.net/images/pdf/LDD3/ch11.pdf
+        * If a user-space program needs to use these types, it can prefix the names
+        * with a double underscore: __u8 and the other types are defined independent
+        * of __KERNEL__. If, for example, a driver needs to exchange binary structures
+        * with a program running in user space by means of ioctl, the header files
+        * should declare 32-bit fields in the structures as __u32.
+        */
+       __u64 timestamp;                                // Monotonically increasing pseudo timestamp
+       unsigned int obj_id     ;                       // The event_id of the object this event is related to
+       unsigned long instr_id;                         // The id of the instruction that triggers this event, e.g., calls kfree
+       unsigned long trace[NR_MAX_TRACE_ENTRIES];      // Stack-trace leading to the event, starting at the syscall
+};
+
+struct ev_rbtree_node {
+       struct rb_node node;    // rb_tree node for the rb_tree interface
+       char *ptr;              // Pointer of the allocation, also used as key for sorting
+       size_t size;            // size of allocation
+       atomic64_t r_timestamp; // timestamp of the most recently recorded read
+       atomic64_t w_timestamp; // timestamp of the most recently recorded write
+};
+
+#endif /* _LINUX_EVTRACK_IOCTLS_H */
+
diff --git a/kernel/Makefile b/kernel/Makefile
index 10ef068f598d..15f08fc17722 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -3,7 +3,7 @@
 # Makefile for the linux kernel.
 #
 
-obj-y     = fork.o exec_domain.o panic.o \
+obj-y     = hypercall.o fork.o exec_domain.o panic.o \
 	    cpu.o exit.o softirq.o resource.o \
 	    sysctl.o capability.o ptrace.o user.o \
 	    signal.o sys.o umh.o workqueue.o pid.o task_work.o \
@@ -38,6 +38,11 @@ KCOV_INSTRUMENT_kcov.o := n
 KASAN_SANITIZE_kcov.o := n
 KCSAN_SANITIZE_kcov.o := n
 UBSAN_SANITIZE_kcov.o := n
+# If EVTRACK is self-instrumented, the instrumentation
+# generates infinite recursive calls to itself -- which
+# makes booting fail at an early stage.
+KCOV_INSTRUMENT_evtrack.o := n
+KASAN_SANITIZE_evtrack.o := n
 KMSAN_SANITIZE_kcov.o := n
 CFLAGS_kcov.o := $(call cc-option, -fno-conserve-stack) -fno-stack-protector
 
@@ -86,6 +91,11 @@ obj-$(CONFIG_AUDIT) += audit.o auditfilter.o
 obj-$(CONFIG_AUDITSYSCALL) += auditsc.o audit_watch.o audit_fsnotify.o audit_tree.o
 obj-$(CONFIG_GCOV_KERNEL) += gcov/
 obj-$(CONFIG_KCOV) += kcov.o
+obj-$(CONFIG_EVTRACK) += evtrack.o
+CFLAGS_evtrack.o := -DDEBUG
+obj-y += my_syscall.o
+obj-y += my_syscall2.o
+obj-y += my_syscall3.o
 obj-$(CONFIG_KPROBES) += kprobes.o
 obj-$(CONFIG_FAIL_FUNCTION) += fail_function.o
 obj-$(CONFIG_KGDB) += debug/
diff --git a/kernel/evtrack.c b/kernel/evtrack.c
new file mode 100644
index 000000000000..2cd9d59bff86
--- /dev/null
+++ b/kernel/evtrack.c
@@ -0,0 +1,749 @@
+// SPDX-License-Identifier: GPL-2.0
+#define pr_fmt(fmt) "evtrack: " fmt
+
+#include <linux/compiler.h>
+#include <linux/errno.h>
+#include <linux/export.h>
+#include <linux/types.h>
+#include <linux/file.h>
+#include <linux/fs.h>
+#include <linux/hashtable.h>
+#include <linux/init.h>
+#include <linux/mm.h>
+#include <linux/preempt.h>
+#include <linux/printk.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/spinlock.h>
+#include <linux/vmalloc.h>
+#include <linux/debugfs.h>
+#include <linux/uaccess.h>
+#include <linux/evtrack.h>
+#include <linux/refcount.h>
+#include <linux/log2.h>
+#include <asm/setup.h>
+#include <linux/stacktrace.h>
+#include <linux/stackdepot.h>
+#include <linux/string.h>
+#include <linux/kallsyms.h>
+#include <linux/atomic.h>
+
+#define evtrack_debug(fmt, ...) pr_debug("%s: " fmt, __func__, ##__VA_ARGS__)
+
+/* Globals */
+static atomic64_t timestamp = ATOMIC64_INIT(0);
+static char *api_names[] = {"__kasan_kmalloc", "kmem_cache_alloc_trace",
+	"kfree", "slab_post_alloc_hook", "kmalloc_array", "kcalloc",
+	"kzalloc", "kmalloc", "slab_free_freelist_hook", "__kasan_slab_free",
+	"__vmalloc_node", "__asan_load1", "__asan_load2", "__asan_load4",
+	"__asan_load8", "__asan_store1", "__asan_store2", "__asan_store4",
+	"__asan_store8", "kmem_cache_free", "__kmalloc",
+	"kmem_cache_alloc_node", "__kmalloc_track_caller", "kvmalloc_node",
+	"kmalloc_node", "__vmalloc_node_range", "__kmalloc_node",
+	"kmem_cache_alloc_node_trace", "kvfree", "check_memory_region",
+	"kmem_cache_alloc"};
+//
+struct api* api_details;
+struct rb_root alloc_root;
+static spinlock_t rbtree_lock;
+
+/* debugfs */
+bool noconsole = true, noshmem = false;
+unsigned long evcnt = 0, max_evcnt = -1;
+unsigned long long evtypes = ((1 << EVTRACK_EVENT_HEAP_ALLOCATION) | (1 << EVTRACK_EVENT_HEAP_DEALLOCATION) |
+							 (1 << EVTRACK_EVENT_HEAP_READ) | (1 << EVTRACK_EVENT_HEAP_WRITE)) >> 1;
+
+static bool check_evtrack_mode(enum evtrack_mode needed_mode, struct task_struct *t)
+{
+	unsigned int mode;
+
+	/*
+	 * We are interested in event tracking as a function of a syscall inputs,
+	 * so we ignore code executed in interrupts.
+	 */
+	if (!in_task())
+		return false;
+	if (rcu_preempt_depth())
+		return false;
+	if (in_softirq())
+		return false;
+	if (in_serving_softirq())
+		return false;
+	if (in_hardirq())
+		return false;
+	if (in_irq())
+		return false;
+	// if (in_atomic())
+	// 	return false;
+	if (irqs_disabled())
+		return false;
+
+	mode = READ_ONCE(t->evtrack_mode);
+	/*
+	 * There is some code that runs in interrupts but for which
+	 * in_interrupt() returns false (e.g. preempt_schedule_irq()).
+	 * READ_ONCE()/barrier() effectively provides load-acquire wrt
+	 * interrupts, there are paired barrier()/WRITE_ONCE() in
+	 * evtrack_start().
+	 */
+	barrier();
+	return mode == needed_mode;
+}
+
+/*
+ * Return true if an event takes place on an object which we
+ * have observed getting allocated. Return False otherwise.
+ */
+static bool is_known_allocation(struct evtrack_event* event, __u64 ts_old) {
+	struct rb_node *node = alloc_root.rb_node;
+	struct ev_rbtree_node *data;
+	char *ptr = (char *) event->ptr;
+	char *alloc_ptr;
+	size_t alloc_size;
+	__u64 tmp, old;
+
+	spin_lock(&rbtree_lock);
+	while (node) {
+		data = rb_entry(node, struct ev_rbtree_node, node);
+		alloc_ptr = data->ptr;
+		alloc_size = data->size;
+		if (ptr < alloc_ptr)
+			node = node->rb_left;
+		else if (ptr >= (alloc_ptr + alloc_size))
+			node = node->rb_right;
+		else {
+			if (event->type == EVTRACK_EVENT_HEAP_READ) {
+				old = (__u64) atomic64_read(&data->r_timestamp);
+				if (ts_old > old || old == 0) {
+					smp_mb__before_atomic();
+					tmp = (__u64) atomic64_cmpxchg(&data->r_timestamp, old, atomic64_read(&timestamp));
+					smp_mb__after_atomic();
+					if (tmp == old) {
+						event->alloc_ptr = alloc_ptr;
+						spin_unlock(&rbtree_lock);
+						return true;
+					} else {
+						event->alloc_ptr = alloc_ptr;
+						spin_unlock(&rbtree_lock);
+						return false;
+					}
+				} else {
+					event->alloc_ptr = alloc_ptr;
+					spin_unlock(&rbtree_lock);
+					return false;
+				}
+				/*
+				if (ts_old > data->r_timestamp || data->r_timestamp == 0) {
+					// this time stamp may not be exactly the one the event will have later on
+					// because another event in a different thread could get this timestamp.
+					// This is not a problem since we will still detect that there has been
+					// a read for this object.
+					data->r_timestamp = atomic64_read(&timestamp);
+					return true;
+				} else {
+					return false;
+				}
+				*/
+			}
+			if (event->type == EVTRACK_EVENT_HEAP_WRITE) {
+				old = (__u64) atomic64_read(&data->w_timestamp);
+				if (ts_old > old || old == 0) {
+					smp_mb__before_atomic();
+					tmp = (__u64) atomic64_cmpxchg(&data->w_timestamp, old, atomic64_read(&timestamp));
+					smp_mb__after_atomic();
+					if (tmp == old) {
+						event->alloc_ptr = alloc_ptr;
+						spin_unlock(&rbtree_lock);
+						return true;
+					} else {
+						event->alloc_ptr = alloc_ptr;
+						spin_unlock(&rbtree_lock);
+						return false;
+					}
+				} else {
+					event->alloc_ptr = alloc_ptr;
+					spin_unlock(&rbtree_lock);
+					return false;
+				}
+				/*
+				if (ts_old > data->w_timestamp || data->w_timestamp == 0) {
+					data->r_timestamp = atomic64_read(&timestamp);
+					return true;
+				} else {
+					return false;
+				}
+				*/
+			}
+			event->alloc_ptr = alloc_ptr;
+			spin_unlock(&rbtree_lock);
+			return true;
+		}
+	}
+	event->alloc_ptr = alloc_ptr;
+	spin_unlock(&rbtree_lock);
+	return false;
+}
+
+/*
+ * Save the allocation in the red-black tree so that we can
+ * search for it later.
+ */
+static void save_allocation(struct evtrack_event* event) {
+	struct ev_rbtree_node *this;
+	struct rb_node **new = &(alloc_root.rb_node), *parent = NULL;
+	struct ev_rbtree_node *data = kzalloc(sizeof(struct ev_rbtree_node),
+			GFP_KERNEL);
+	if (data == NULL)
+		return;
+	data->ptr = (char *) event->ptr;
+	data->size = event->size;
+	data->r_timestamp = (atomic64_t) ATOMIC64_INIT(0);
+	data->w_timestamp = (atomic64_t) ATOMIC64_INIT(0);
+
+	/* Figure out where to put new node. */
+	spin_lock(&rbtree_lock);
+	while(*new) {
+		this = rb_entry(*new, struct ev_rbtree_node, node);
+		parent = *new;
+		if (this->ptr > data->ptr)
+			new = &((*new)->rb_left);
+		else if ((this->ptr + this->size) <= data->ptr)
+			new = &((*new)->rb_right);
+		else {
+			if (data->size > this->size)
+				this->size = data->size;
+			kfree(data);
+			spin_unlock(&rbtree_lock);
+			return;
+		}
+	}
+
+	/*Add new node and rebalance tree. */
+	rb_link_node(&data->node, parent, new);
+	rb_insert_color(&data->node, &alloc_root);
+	spin_unlock(&rbtree_lock);
+}
+
+/*
+* Identify the instruction that has triggered this event.
+*/
+/*
+static void identify_trigger(struct evtrack_event* event) {
+	unsigned int trace_idx, api_idx, num_api_methods;
+	unsigned long trace_addr;
+	bool is_found, is_not_found;
+	struct api* api_info;
+
+	event->instr_id = 0;
+	is_found = false;
+	num_api_methods = sizeof(api_names) / sizeof(api_names[0]);
+	// Iterate over the call-stack
+	for (trace_idx = 0; trace_idx < event->nr_trace; trace_idx++) {
+		trace_addr = event->trace[trace_idx];
+		is_not_found = true;
+		// Check if a call-stack entry belongs to the
+		// list of kernel API associated with the event
+		for (api_idx = 0; api_idx < num_api_methods; api_idx++) {
+			api_info = api_details + api_idx;
+			if ((trace_addr >= api_info->start) && (trace_addr < (api_info->start + api_info->size))) {
+				is_found = true;
+				is_not_found = false;
+				break;
+			}
+		}
+
+		if (is_found && is_not_found) {
+			event->instr_id = trace_addr;
+			is_found = false;
+		}
+	}
+}
+*/
+
+/*
+* Saves the stack-trace of the event, and filters out a trace
+* if it is IRQ generated. In fact, it returns the depth of the
+* stack-trace until the IRQ caller. If an IRQ caller exists,
+* the returned depth will be smaller than the actual depth.
+*/
+static unsigned int save_stack_trace(struct evtrack_event* event) {
+	unsigned int nr_trace, nr_filtered_trace;
+
+	nr_trace = stack_trace_save(event->trace, NR_MAX_TRACE_ENTRIES, 0);
+	event->nr_trace = nr_trace;
+	nr_filtered_trace = filter_irq_stacks(event->trace, nr_trace);
+	return nr_filtered_trace;
+}
+
+/*
+* Filter out types of events not being tracked.
+* Specific event types can be enabled or disabled
+* by writing to /sys/kernel/debug/ev/evcnt/evtypes
+* debugfs file.
+*/
+
+void record_event(enum evtrack_event_types event_type, const void* ptr,
+                    size_t size, bool keep_tag) {
+	struct task_struct *t;
+	struct evtrack_event *events, *event;
+	unsigned int evtrack_size, nr_filtered_trace;
+	unsigned long nr_events;
+
+	/*
+	 * We are interested in event tracking as a function of a syscall inputs,
+	 * so we ignore code executed in interrupts.
+	 */
+	t = current;
+	if (!check_evtrack_mode(EVTRACK_MODE_TRACK_ALL, t)) {
+		return;
+	}
+
+	/*
+	 * We must disable EVTRACK here to avoid recursive calls
+	 * from the instrumented kernel APIs that we use below.
+	 * Recursive EVTRACK calls from inside the APIs not only
+	 * generate irrelevant events, but also blows up the kernel
+	 * stack. The root cause of the issue is hard to pin down,
+	 * as it is manifested in many forms, e.g., deadlock, memory
+	 * corruptions, etc., with varying contexts in different executions.
+	 */
+	WRITE_ONCE(t->evtrack_mode, EVTRACK_MODE_DISABLED);
+
+	events = (struct evtrack_event*)t->evtrack_area;
+	event = events;
+	/* The size field of the first entry contains the number of events recorded. */
+	nr_events = READ_ONCE(events->size) + 1;
+	evtrack_size = nr_events * sizeof(struct evtrack_event);
+	if (likely(evtrack_size < t->evtrack_size)) {
+		event += nr_events;
+
+		// Filter the event out if it is not of a type being tracked
+		if (!(evtypes & (1ul << (event_type - 1))))
+			goto out;
+
+		// Filter out a READ/WRITE event if the allocation hasn't been seen before
+		event->ptr = ptr;
+		event->instr_id = 0;
+		event->size = size;
+		event->type = event_type;
+		if (event_type == EVTRACK_EVENT_HEAP_READ
+				|| event_type == EVTRACK_EVENT_HEAP_WRITE
+				|| event_type == EVTRACK_EVENT_HEAP_DEALLOCATION) {
+			if (!is_known_allocation(event, t->evtrack->timestamp))
+			{
+				// if(event_type == 2){
+				// 	printk(KERN_INFO "[PID: %d] record_event:  unknown_allocation: type: %d, addr: %p, size: %d, alloc: %p\n", current->pid, event_type, (void*)ptr, size, (void*)event->alloc_ptr);
+				// }
+				goto out;
+			}
+				
+		} else {
+			// insert event into rb tree
+			save_allocation(event);
+			event->obj_id = 0;				// EVTRACK_EVENT_HEAP_ALLOCATION
+		}
+		// if(event_type < 3){
+		// 	printk(KERN_INFO "[PID: %d] record_event: type: %d, addr: %p, size: %d, alloc: %p\n", current->pid, event_type, (void*)ptr, size, (void*)event->alloc_ptr);
+		// }
+		
+
+		// Filter out an event if generated by an IRQ caller
+		nr_filtered_trace = save_stack_trace(event);
+		if (nr_filtered_trace < event->nr_trace)
+			goto out;
+		evcnt += 1;
+		event->event_id = nr_events;
+		//event->type = event_type;
+		// event->timestamp = ktime_get_ns();
+		event->timestamp = (__u64) atomic64_inc_return(&timestamp) - 1;
+		//event->timestamp = ++timestamp;		// A monotonically increasing pseudo timestamp
+
+		// Identify the instruction that has triggered this event
+		//identify_trigger(event);
+
+		if (!noshmem)
+			WRITE_ONCE(events->size, nr_events);
+
+	}
+
+	if (evcnt >= max_evcnt)
+		return;
+
+out:
+	// Re-enable EVTRACK that we disabled above (in the beginning)
+	WRITE_ONCE(t->evtrack_mode, EVTRACK_MODE_TRACK_ALL);
+}
+
+static void evtrack_start(struct task_struct *t, unsigned int size,
+		void *area, enum evtrack_mode mode)
+{
+	evtrack_debug("t=%px, size=%u, area=%px\n", t, size, area);
+	/* Cache in task struct for performance. */
+	t->evtrack_size = size;
+	t->evtrack_area = area;
+	/* See comment in check_evtrack_mode(). */
+	barrier();
+	WRITE_ONCE(t->evtrack_mode, mode);
+}
+
+static void evtrack_restart(struct task_struct *t, struct evtrack *evtrack)
+{
+	WRITE_ONCE(t->evtrack_mode, EVTRACK_MODE_TRACK_ALL);
+	barrier();
+	evtrack->mode = EVTRACK_MODE_TRACK_ALL;
+}
+
+static void evtrack_stop(struct task_struct *t, struct evtrack *evtrack)
+{
+	WRITE_ONCE(t->evtrack_mode, EVTRACK_MODE_STOP);
+	barrier();
+	evtrack->mode = EVTRACK_MODE_STOP;
+}
+
+static void evtrack_task_reset(struct task_struct *t)
+{
+	WRITE_ONCE(t->evtrack_mode, EVTRACK_MODE_DISABLED);
+	barrier();
+	t->evtrack_size = 0;
+	t->evtrack_area = NULL;
+	t->evtrack = NULL;
+}
+
+void evtrack_task_init(struct task_struct *t)
+{
+	evtrack_task_reset(t);
+}
+
+static void evtrack_reset(struct evtrack *evtrack)
+{
+	evtrack->t = NULL;
+	evtrack->mode = EVTRACK_MODE_INIT;
+}
+
+static void evtrack_disable(struct task_struct *t, struct evtrack *evtrack)
+{
+	evtrack_task_reset(t);
+	evtrack_reset(evtrack);
+}
+
+static void evtrack_get(struct evtrack *evtrack)
+{
+	refcount_inc(&evtrack->refcount);
+}
+
+static void free_rb_tree_node(struct rb_node *node) {
+	struct ev_rbtree_node *data;
+	if (!node)
+		return;
+	if (node->rb_right)
+		free_rb_tree_node(node->rb_right);
+	if (node->rb_left)
+		free_rb_tree_node(node->rb_left);
+	data = rb_entry(node, struct ev_rbtree_node, node);
+	kfree(data);
+}
+
+static void free_rb_tree(struct evtrack *evtrack) {
+	struct rb_root *root = &(evtrack->root);
+	free_rb_tree_node(root->rb_node);
+	root->rb_node = NULL;
+}
+
+static void evtrack_put(struct evtrack *evtrack)
+{
+	if (refcount_dec_and_test(&evtrack->refcount)) {
+		vfree(evtrack->area);
+		free_rb_tree(evtrack);
+		kfree(evtrack);
+	}
+}
+
+void evtrack_task_exit(struct task_struct *t)
+{
+	struct evtrack *evtrack;
+
+	evtrack = t->evtrack;
+	if (evtrack == NULL)
+		return;
+
+	spin_lock(&evtrack->lock);
+	evtrack_debug("t=%px, evtrack->t=%px\n", t, evtrack->t);
+	/*
+	 * For EVTRACK_ENABLE devices we want to make sure that t->evtrack->t == t,
+	 * which comes down to:
+	 *        WARN_ON(!evtrack->remote && evtrack->t != t);
+	 *
+	 * 2. The task that created evtrack exiting without calling EVTRACK_DISABLE,
+	 *    and then again we can make sure that t->evtrack->t == t:
+	 *        WARN_ON(evtrack->remote && evtrack->t != t);
+	 *
+	 * By combining all three checks into one we get:
+	 */
+	if (WARN_ON(evtrack->t != t)) {
+		spin_unlock(&evtrack->lock);
+		return;
+	}
+	/* Just to not leave dangling references behind. */
+	evtrack_disable(t, evtrack);
+	spin_unlock(&evtrack->lock);
+	evtrack_put(evtrack);
+	// TODO if nobody called ioctl(disable) before the task exits and nobody closed the file, then the refcount will
+	// stay at 1 right?
+}
+
+static int evtrack_mmap(struct file *filep, struct vm_area_struct *vma)
+{
+	int res = 0;
+	void *area;
+	struct evtrack *evtrack = vma->vm_file->private_data;
+	unsigned long size, off;
+	struct page *page;
+
+	area = vmalloc_user(vma->vm_end - vma->vm_start);
+	if (!area)
+		return -ENOMEM;
+
+	spin_lock(&evtrack->lock);
+	size = evtrack->nr_events * sizeof(struct evtrack_event);
+	if (evtrack->mode != EVTRACK_MODE_INIT || vma->vm_pgoff != 0 ||
+			vma->vm_end - vma->vm_start != size) {
+		res = -EINVAL;
+		goto exit;
+	}
+	if (!evtrack->area) {
+		evtrack->area = area;
+		vma->vm_flags |= VM_DONTEXPAND;
+		spin_unlock(&evtrack->lock);
+		for (off = 0; off < size; off += PAGE_SIZE) {
+			page = vmalloc_to_page(evtrack->area + off);
+			if (vm_insert_page(vma, vma->vm_start + off, page))
+				WARN_ONCE(1, "vm_insert_page() failed");
+		}
+		return 0;
+	}
+exit:
+	spin_unlock(&evtrack->lock);
+	vfree(area);
+	return res;
+}
+
+static int evtrack_open(struct inode *inode, struct file *filep)
+{
+	struct evtrack *evtrack;
+
+	evtrack = kzalloc(sizeof(*evtrack), GFP_KERNEL);
+	if (!evtrack)
+		return -ENOMEM;
+	evtrack->mode = EVTRACK_MODE_DISABLED;
+	/*
+	 * refcount is initially set to 1 to enforce cleaning
+	 * up of the resources after (1) EVTRACK_DISABLE ioctl call
+	 * and (2) close(fd), or task exit. Both (1) and (2) have
+	 * to take place, otherwise there will be a memory leak.
+	 */
+	refcount_set(&evtrack->refcount, 1);
+	spin_lock_init(&evtrack->lock);
+	filep->private_data = evtrack;
+	return nonseekable_open(inode, filep);
+}
+
+static int evtrack_close(struct inode *inode, struct file *filep)
+{
+	evtrack_put(filep->private_data);
+	return 0;
+}
+
+/*
+* Translates user evtrack mode to the internal one.
+*/
+static int evtrack_get_mode(unsigned long arg)
+{
+	if (arg == EVTRACK_TRACK_ALL)
+		return EVTRACK_MODE_TRACK_ALL;
+	else
+		return -EINVAL;
+}
+
+static int evtrack_ioctl_locked(struct evtrack *evtrack, unsigned int cmd,
+                 unsigned long arg)
+{
+	struct task_struct *t = current;
+	unsigned long nr_events, unused;
+	unsigned int size;
+	int mode, prev_mode;
+
+	// DEBUG
+	unsigned int dbg_code, i;
+	unsigned long symbolsize, offset, num_api_methods;
+	struct api* api_info;
+
+
+	switch (cmd) {
+		case EVTRACK_INIT_TRACK:
+			evtrack_debug("EVTRACK_INIT_TRACK (pid=%d, %s)\n", t->pid, t->comm);
+			/*
+			* Enable evtrack in trace mode and setup buffer size.
+			* Must happen before anything else.
+			*/
+			if (evtrack->mode != EVTRACK_MODE_DISABLED)
+				return -EBUSY;
+			/*
+			* Size must be at least 2 to hold current position and one PC.
+			* Later we allocate nr_events * (sizeof(struct evtrack_event)
+			* bytes of memory that must not overflow.
+			*/
+			nr_events = arg;
+			if (nr_events < 2 || nr_events > INT_MAX / sizeof(struct evtrack_event))
+				return -EINVAL;
+			evtrack->nr_events = nr_events;
+			evtrack->mode = EVTRACK_MODE_INIT;
+			return 0;
+
+		case EVTRACK_ENABLE:
+			evtrack_debug("EVTRACK_ENABLE (pid=%d, %s)\n", t->pid, t->comm);
+			/*
+			* Enable event tracking for the current task.
+			* At this point user must have been enabled tracking mode,
+			* and mmapped the file. Event tracking is disabled only
+			* at task exit or voluntary by EVTRACK_DISABLE. After that it can
+			* be enabled for another task.
+			*/
+			if ((evtrack->mode != EVTRACK_MODE_INIT && evtrack->mode != EVTRACK_MODE_STOP) || !evtrack->area)
+				return -EINVAL;
+			if (evtrack->mode == EVTRACK_MODE_INIT && (evtrack->t != NULL || t->evtrack != NULL))
+				return -EBUSY;
+			prev_mode = evtrack->mode;
+			mode = evtrack_get_mode(arg);
+			if (mode < 0)
+				return mode;
+			evtrack->mode = mode;
+			evtrack->timestamp = (__u64) atomic64_read(&timestamp);
+			if (prev_mode == EVTRACK_MODE_STOP) {
+				evtrack_restart(t, evtrack);
+				return 0;
+			}
+			size = evtrack->nr_events * sizeof(struct evtrack_event);
+			evtrack_start(t, size, evtrack->area, evtrack->mode);
+			t->evtrack = evtrack;
+			evtrack->t = t;
+			evtrack->root = RB_ROOT;
+			/* Put either in evtrack_task_exit() or in EVTRACK_DISABLE. */
+			evtrack_get(evtrack);
+			return 0;
+
+		case EVTRACK_STOP:
+			evtrack_debug("EVTRACK_STOP (pid=%d, %s)\n", t->pid, t->comm);
+			/* temporarily stop tracing without deleting any important structures */
+			unused = arg;
+			if (unused != 0 || current->evtrack != evtrack)
+				return -EINVAL;
+			if (WARN_ON(evtrack->t != t))
+				return -EINVAL;
+			evtrack_stop(t, evtrack);
+			return 0;
+
+		case EVTRACK_DISABLE:
+			evtrack_debug("EVTRACK_DISABLE (pid=%d, %s)\n", t->pid, t->comm);
+			/* Disable event tracking for the current task. */
+			unused = arg;
+			if (unused != 0 || current->evtrack != evtrack)
+				return -EINVAL;
+			if (WARN_ON(evtrack->t != t))
+				return -EINVAL;
+			evtrack_disable(t, evtrack);
+			evtrack_put(evtrack);
+			return 0;
+
+		case EVTRACK_DEBUG:
+			dbg_code = arg;
+			switch (dbg_code) {
+				case 1:
+					kallsyms_lookup_size_offset((unsigned long)evtrack_task_init, &symbolsize, &offset);
+					evtrack_debug("EVTRACK_DEBUG %px, %ld, %ld\n", evtrack_task_init, symbolsize, offset);
+					return 0;
+
+				case 2:
+					evtrack_debug("EVTRACK_DEBUG 0x%lx\n", kallsyms_lookup_name("evtrack_task_init"));
+					return 0;
+
+				case 3:
+					api_info = api_details;
+					num_api_methods = sizeof(api_names) / sizeof(api_names[0]);
+					for (i = 0; i < num_api_methods; i++) {
+						evtrack_debug("EVTRACK_DEBUG %s, 0x%lx, 0x%lx\n", api_info->name, api_info->start, api_info->size);
+					}
+					return 0;
+			}
+			return 1;
+
+		default:
+			return -ENOTTY;
+	}
+
+}
+
+static long evtrack_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
+{
+    struct evtrack *evtrack;
+	int res;
+
+	evtrack = filep->private_data;
+	spin_lock(&evtrack->lock);
+	res = evtrack_ioctl_locked(evtrack, cmd, arg);
+	spin_unlock(&evtrack->lock);
+
+	return res;
+}
+
+static const struct file_operations evtrack_fops = {
+    .open		    = evtrack_open,
+    .unlocked_ioctl	= evtrack_ioctl,
+    .compat_ioctl	= evtrack_ioctl,
+    .mmap           = evtrack_mmap,
+    .release        = evtrack_close,
+};
+
+void populate_api_info(void) {
+	struct api* api_info;
+	char *api_name;
+	unsigned long i, addr, symbolsize, offset, num_api_methods;
+
+	api_details = (struct api*)kmalloc(sizeof(struct api) * sizeof(api_names), GFP_KERNEL);
+	api_info = api_details;
+	num_api_methods = sizeof(api_names) / sizeof(api_names[0]);
+	for (i = 0; i < num_api_methods; i++) {
+		api_name = api_names[i];
+		addr = kallsyms_lookup_name(api_name);
+		if (addr) {
+			kallsyms_lookup_size_offset(addr, &symbolsize, &offset);
+			strncpy(api_info->name, api_name, 256);
+			api_info->start = addr;
+			api_info->size = symbolsize;
+			api_info++;
+		}
+	}
+}
+
+static int __init evtrack_init(void)
+{
+	struct dentry* evtrack_dir;
+
+    /*
+	 * The EVTRACK debugfs file won't ever get removed and thus,
+	 * there is no need to protect it against removal races. The
+	 * use of debugfs_create_file_unsafe() is actually safe here.
+	 */
+    debugfs_create_file_unsafe("evtrack", 0600, NULL, NULL, &evtrack_fops);
+	// Ref: https://www.kernel.org/doc/html/latest/filesystems/debugfs.html
+	evtrack_dir = debugfs_create_dir("ev", NULL);
+	debugfs_create_bool("noconsole", 0600, evtrack_dir, &noconsole);
+	debugfs_create_bool("noshmem", 0600, evtrack_dir, &noshmem);
+	debugfs_create_ulong("evcnt", 0600, evtrack_dir, &evcnt);
+	debugfs_create_ulong("max_evcnt", 0600, evtrack_dir, &max_evcnt);
+	debugfs_create_x64("evtypes", 0600, evtrack_dir, &evtypes);
+
+	populate_api_info();
+    return 0;
+}
+
+device_initcall(evtrack_init);
diff --git a/kernel/exit.c b/kernel/exit.c
index 15dc2ec80c46..627d8a1855c1 100644
--- a/kernel/exit.c
+++ b/kernel/exit.c
@@ -60,6 +60,7 @@
 #include <linux/writeback.h>
 #include <linux/shm.h>
 #include <linux/kcov.h>
+#include <linux/evtrack.h>
 #include <linux/kmsan.h>
 #include <linux/random.h>
 #include <linux/rcuwait.h>
@@ -812,6 +813,7 @@ void __noreturn do_exit(long code)
 	WARN_ON(tsk->plug);
 
 	kcov_task_exit(tsk);
+	evtrack_task_exit(tsk);
 	kmsan_task_exit(tsk);
 
 	coredump_task_exit(tsk);
diff --git a/kernel/fork.c b/kernel/fork.c
index 9f7fe3541897..00960f257d33 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -109,6 +109,10 @@
 #define CREATE_TRACE_POINTS
 #include <trace/events/task.h>
 
+#ifdef CONFIG_EVTRACK
+#include <linux/evtrack.h>
+#endif
+
 /*
  * Minimum number of threads to boot the kernel
  */
@@ -997,6 +1001,10 @@ static struct task_struct *dup_task_struct(struct task_struct *orig, int node)
 	if (err)
 		goto free_stack;
 
+#ifdef CONFIG_EVTRACK
+	evtrack_task_init(tsk);
+#endif
+
 #ifdef CONFIG_SECCOMP
 	/*
 	 * We must handle setting up seccomp filters once we're under
diff --git a/kernel/hypercall.c b/kernel/hypercall.c
new file mode 100644
index 000000000000..65a8334ae7fa
--- /dev/null
+++ b/kernel/hypercall.c
@@ -0,0 +1,33 @@
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/sched.h>
+#include <linux/syscalls.h>
+
+SYSCALL_DEFINE4(hypercall, unsigned long, id, unsigned long, cmd, unsigned long, arg, unsigned long, sched)
+{
+    uint64_t ret;
+    asm volatile (
+            /*
+            "pushq %%rax\n\t"
+            "pushq %%rbx\n\t"
+            "pushq %%rcx\n\t"
+            "pushq %%rdx\n\t"
+            */
+            "movq %1, %%rax\n\t"
+            "movq %2, %%rbx\n\t"
+            "movq %3, %%rcx\n\t"
+            "movq %4, %%rdx\n\t"
+            "nop\n\t"
+            "movq %%rax, %0\n\t"
+            /*
+            "popq %%rdx\n\t"
+            "popq %%rcx\n\t"
+            "popq %%rbx\n\t"
+            "popq %%rax\n\t"
+            */
+            :"=r"(ret)
+            :"r"(id), "r"(cmd), "r"(arg), "r"(sched)
+            :
+            );
+    return ret;
+}
\ No newline at end of file
diff --git a/kernel/my_syscall.c b/kernel/my_syscall.c
new file mode 100644
index 000000000000..dea0ea14d9a3
--- /dev/null
+++ b/kernel/my_syscall.c
@@ -0,0 +1,55 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+
+/*
+* Using the the following pragmas to prevent gcc optimize
+* out `ptr` had the unintended side effect of printk()
+* not printing raw kernel addresses any more. If the
+* pramgas are used, dmesg contains the following:
+* sys_mycall: ptr = 0000000000000010
+* Ref: https://github.com/torvalds/linux/blob/master/Documentation/core-api/printk-formats.rst
+* According to the doc above, "%px" is supposed to
+* print unmodified kernel address, albeit unsafe.
+* To workaround the gcc optimization issue, we
+* mark the pointer as `volatile`
+*/
+
+// #pragma GCC push_options
+// #pragma GCC optimize("O0")
+
+#define NR_MAX_TRACE_ENTRIES 32
+
+SYSCALL_DEFINE0(mycall)
+{
+    volatile char* heap_ptr;
+    volatile unsigned long* heap_write_ptr;
+    volatile unsigned long stack_var;
+    /*
+    unsigned int nr_trace, i;
+    unsigned long trace[NR_MAX_TRACE_ENTRIES];
+    */
+
+    /* printk("sys_mycall: stack-trace\n");
+    // Print stack-trace
+    nr_trace = stack_trace_save(trace, NR_MAX_TRACE_ENTRIES, 0);
+    nr_trace = (nr_trace <= NR_MAX_TRACE_ENTRIES) ? nr_trace : NR_MAX_TRACE_ENTRIES;
+    for(i = 0; i < nr_trace; i++)
+        printk("|----> (%d) %lx\n", i, trace[i]); */
+
+    heap_ptr = (char*)kmalloc(2048, GFP_KERNEL);
+    // printk("sys_mycall: kmalloc=%px\n", heap_ptr);
+
+    // Test if __asan_loadN/__asan_storeN instruments both stack and heap load/store
+    stack_var = 0xcafebabe;
+    heap_write_ptr = (unsigned long*)(heap_ptr + 0x10);
+    *heap_write_ptr = 0xdeadbeef;
+    printk("sys_mycall: stack_write=%px, heap_write=%px\n", &stack_var, heap_write_ptr);
+
+    // Casting was necessary to suppress kfree() warning about volatile pointer
+    kfree((void*)heap_ptr);
+    // printk("sys_mycall: kfree=%px\n", heap_ptr);
+
+    return 0;
+}
+
+// #pragma GCC pop_options
diff --git a/kernel/my_syscall2.c b/kernel/my_syscall2.c
new file mode 100644
index 000000000000..8439a46bf3ed
--- /dev/null
+++ b/kernel/my_syscall2.c
@@ -0,0 +1,42 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+#include <linux/mycall.h>
+
+/*
+ * Using the the following pragmas to prevent gcc optimize
+ * out `ptr` had the unintended side effect of printk()
+ * not printing raw kernel addresses any more. If the
+ * pramgas are used, dmesg contains the following:
+ * sys_mycall: ptr = 0000000000000010
+ * Ref: https://github.com/torvalds/linux/blob/master/Documentation/core-api/printk-formats.rst
+ * According to the doc above, "%px" is supposed to
+ * print unmodified kernel address, albeit unsafe.
+ * To workaround the gcc optimization issue, we
+ * mark the pointer as `volatile`
+ */
+
+// #pragma GCC push_options
+// #pragma GCC optimize("O0")
+
+#define NR_MAX_TRACE_ENTRIES 32
+
+static char *heap_ptr;
+
+SYSCALL_DEFINE0(mycall2)
+{
+	volatile unsigned long* heap_write_ptr;
+
+	heap_ptr = (char*)kmalloc(32, GFP_KERNEL);
+
+	heap_write_ptr = (unsigned long*)(heap_ptr + 0x10);
+	*heap_write_ptr = 0xdeadbeef;
+	printk("sys_mycall2: wrote value %lx\n", *heap_write_ptr);
+
+	return 0;
+}
+
+char *get_my_pointer(void) {
+	return heap_ptr;
+}
+
+// #pragma GCC pop_options
diff --git a/kernel/my_syscall3.c b/kernel/my_syscall3.c
new file mode 100644
index 000000000000..f928266bd9e8
--- /dev/null
+++ b/kernel/my_syscall3.c
@@ -0,0 +1,39 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+#include <linux/mycall.h>
+
+/*
+ * Using the the following pragmas to prevent gcc optimize
+ * out `ptr` had the unintended side effect of printk()
+ * not printing raw kernel addresses any more. If the
+ * pramgas are used, dmesg contains the following:
+ * sys_mycall: ptr = 0000000000000010
+ * Ref: https://github.com/torvalds/linux/blob/master/Documentation/core-api/printk-formats.rst
+ * According to the doc above, "%px" is supposed to
+ * print unmodified kernel address, albeit unsafe.
+ * To workaround the gcc optimization issue, we
+ * mark the pointer as `volatile`
+ */
+
+// #pragma GCC push_options
+// #pragma GCC optimize("O0")
+
+#define NR_MAX_TRACE_ENTRIES 32
+
+SYSCALL_DEFINE0(mycall3)
+{
+
+	volatile unsigned long* heap_write_ptr;
+	volatile unsigned long my_val;
+	volatile char *heap_ptr = get_my_pointer();
+	printk("sys_mycall3: get_my_pointer(): %px\n", heap_ptr);
+	heap_write_ptr = (unsigned long*)(heap_ptr + 0x10);
+	my_val = *heap_write_ptr;
+	printk("sys_mycall3: read value %lx\n", my_val);
+	*heap_write_ptr = 0xcafebabe;
+	printk("sys_mycall3: new value %lx\n", *heap_write_ptr);
+
+	kfree((void*)heap_ptr);
+	return 0;
+}
+// #pragma GCC pop_options
diff --git a/lib/Kconfig.kasan b/lib/Kconfig.kasan
index be6ee6020290..890e9d5edee2 100644
--- a/lib/Kconfig.kasan
+++ b/lib/Kconfig.kasan
@@ -25,6 +25,15 @@ config CC_HAS_KASAN_GENERIC
 config CC_HAS_KASAN_SW_TAGS
 	def_bool $(cc-option, -fsanitize=kernel-hwaddress)
 
+config EVTRACK
+	bool "EVTRACK: User-space tracking of kernel events"
+	depends on KASAN
+	help
+	  Track kernel-space events, e.g. memory allocation,
+	  deallocation, read/write of stack/heap/static variables,
+	  copying of data to user-space etc.
+
+
 # This option is only required for software KASAN modes.
 # Old GCC versions do not have proper support for no_sanitize_address.
 # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89124 for details.
diff --git a/mm/kasan/common.c b/mm/kasan/common.c
index 21e66d7f261d..1124bd97032a 100644
--- a/mm/kasan/common.c
+++ b/mm/kasan/common.c
@@ -26,6 +26,7 @@
 #include <linux/string.h>
 #include <linux/types.h>
 #include <linux/bug.h>
+#include <linux/evtrack.h>
 
 #include "kasan.h"
 #include "../slab.h"
@@ -205,6 +206,10 @@ static inline bool ____kasan_slab_free(struct kmem_cache *cache, void *object,
 	if (!kasan_arch_is_ready())
 		return false;
 
+#ifdef CONFIG_EVTRACK
+	record_event(EVTRACK_EVENT_HEAP_DEALLOCATION, object, 0, 0);
+#endif
+
 	tagged_object = object;
 	object = kasan_reset_tag(object);
 
@@ -326,6 +331,11 @@ void * __must_check __kasan_slab_alloc(struct kmem_cache *cache,
 	/* Save alloc info (if possible) for non-kmalloc() allocations. */
 	if (kasan_stack_collection_enabled() && !cache->kasan_info.is_kmalloc)
 		kasan_save_alloc_info(cache, tagged_object, flags);
+	
+#ifdef CONFIG_EVTRACK
+	//keep_tag is not used
+	record_event(EVTRACK_EVENT_HEAP_ALLOCATION, tagged_object, cache->object_size, 0);
+#endif
 
 	return tagged_object;
 }
@@ -336,6 +346,12 @@ static inline void *____kasan_kmalloc(struct kmem_cache *cache,
 	unsigned long redzone_start;
 	unsigned long redzone_end;
 
+#ifdef CONFIG_EVTRACK
+	//keep_tag is not used
+	record_event(EVTRACK_EVENT_HEAP_ALLOCATION, object, size, 0);
+#endif
+
+
 	if (gfpflags_allow_blocking(flags))
 		kasan_quarantine_reduce();
 
diff --git a/mm/kasan/generic.c b/mm/kasan/generic.c
index cb762982c8ba..71e92d306614 100644
--- a/mm/kasan/generic.c
+++ b/mm/kasan/generic.c
@@ -30,6 +30,7 @@
 #include <linux/types.h>
 #include <linux/vmalloc.h>
 #include <linux/bug.h>
+#include <linux/evtrack.h>
 
 #include "kasan.h"
 #include "../slab.h"
@@ -177,6 +178,13 @@ static __always_inline bool check_region_inline(unsigned long addr,
 		return !kasan_report(addr, size, write, ret_ip);
 	}
 
+#ifdef CONFIG_EVTRACK
+	if (write)
+		record_event(EVTRACK_EVENT_HEAP_WRITE, (void*)addr, size, write);
+	else
+		record_event(EVTRACK_EVENT_HEAP_READ, (void*)addr, size, write);
+#endif
+
 	if (likely(!memory_is_poisoned(addr, size)))
 		return true;
 
diff --git a/net/unix/af_unix.c b/net/unix/af_unix.c
index f0c2293f1d3b..942a8d1ca9a6 100644
--- a/net/unix/af_unix.c
+++ b/net/unix/af_unix.c
@@ -115,6 +115,7 @@
 #include <linux/freezer.h>
 #include <linux/file.h>
 #include <linux/btf_ids.h>
+#include <linux/delay.h>
 
 #include "scm.h"
 
@@ -2111,6 +2112,7 @@ static int queue_oob(struct socket *sock, struct msghdr *msg, struct sock *other
 	int err = 0;
 
 	skb = sock_alloc_send_skb(sock->sk, 1, msg->msg_flags & MSG_DONTWAIT, &err);
+	// printk(KERN_INFO "[PID: %d] queue_oob: alloc new skb: %p\n", current->pid, (void*)skb);
 
 	if (!skb)
 		return err;
@@ -2124,6 +2126,7 @@ static int queue_oob(struct socket *sock, struct msghdr *msg, struct sock *other
 	}
 
 	unix_state_lock(other);
+	// printk(KERN_INFO "[PID: %d] queue_oob: get lock: %p\n", current->pid, (void*)other);
 
 	if (sock_flag(other, SOCK_DEAD) ||
 	    (other->sk_shutdown & RCV_SHUTDOWN)) {
@@ -2135,14 +2138,23 @@ static int queue_oob(struct socket *sock, struct msghdr *msg, struct sock *other
 	maybe_add_creds(skb, sock, other);
 	skb_get(skb);
 
-	if (ousk->oob_skb)
-		consume_skb(ousk->oob_skb);
+	// printk(KERN_INFO "[PID: %d] queue_oob: free old oob_skb: %p\n", current->pid, (void*)ousk->oob_skb);
+	if (ousk->oob_skb){
+
+		// printk(KERN_INFO "[PID: %d] queue_oob: old oob_skb: %p, users: %d\n", current->pid, (void*)ousk->oob_skb, refcount_read(&ousk->oob_skb->users));
+		if (skb_unref(ousk->oob_skb)){
+			// printk(KERN_INFO "[PID: %d] consume_skb: free old oob_skb: %p, users: %d\n", current->pid, (void*)ousk->oob_skb, refcount_read(&ousk->oob_skb->users));
+			__kfree_skb(ousk->oob_skb);
+		}
+	}
+		// consume_skb(ousk->oob_skb);
 
 	WRITE_ONCE(ousk->oob_skb, skb);
 
 	scm_stat_add(other, skb);
 	skb_queue_tail(&other->sk_receive_queue, skb);
 	sk_send_sigurg(other);
+	// printk(KERN_INFO "[PID: %d] queue_oob: release lock: %p\n", current->pid, (void*)other);
 	unix_state_unlock(other);
 	other->sk_data_ready(other);
 
@@ -2636,6 +2648,7 @@ static int unix_stream_recv_urg(struct unix_stream_read_state *state)
 	}
 
 	oob_skb = u->oob_skb;
+	// printk(KERN_INFO "[PID: %d]  unix_stream_recv_urg: oob_skb: %p, users:%d \n", current->pid, (void*)oob_skb, refcount_read(&oob_skb->users));
 
 	if (!(state->flags & MSG_PEEK))
 		WRITE_ONCE(u->oob_skb, NULL);
@@ -2646,6 +2659,7 @@ static int unix_stream_recv_urg(struct unix_stream_read_state *state)
 
 	if (!(state->flags & MSG_PEEK)) {
 		UNIXCB(oob_skb).consumed += 1;
+		// printk(KERN_INFO "[PID: %d]  unix_stream_recv_urg: kfree_skb oob_skb: %p, users: %d\n", current->pid, (void*)oob_skb, refcount_read(&oob_skb->users));
 		kfree_skb(oob_skb);
 	}
 
@@ -2665,18 +2679,22 @@ static struct sk_buff *manage_oob(struct sk_buff *skb, struct sock *sk,
 
 	if (!unix_skb_len(skb) && !(flags & MSG_PEEK)) {
 		skb_unlink(skb, &sk->sk_receive_queue);
+		// printk(KERN_INFO "[PID: %d]  manage_oob: len=consumed skb: %p, users: %d, len: %d, consumed: %d\n", current->pid, (void*)skb, refcount_read(&skb->users), skb->len, UNIXCB(skb).consumed);
 		consume_skb(skb);
 		skb = NULL;
 	} else {
 		if (skb == u->oob_skb) {
 			if (copied) {
+				// printk(KERN_INFO "[PID: %d]  manage_oob: NULL skb: %p, users: %d\n", current->pid, (void*)skb, refcount_read(&skb->users));
 				skb = NULL;
 			} else if (sock_flag(sk, SOCK_URGINLINE)) {
 				if (!(flags & MSG_PEEK)) {
+					// printk(KERN_INFO "[PID: %d]  manage_oob: SOCK_URGINLINE skb: %p, users: %d, len: %d, consumed: %d\n", current->pid, (void*)skb, refcount_read(&skb->users), skb->len, UNIXCB(skb).consumed);
 					WRITE_ONCE(u->oob_skb, NULL);
 					consume_skb(skb);
 				}
 			} else if (!(flags & MSG_PEEK)) {
+				// printk(KERN_INFO "[PID: %d]  manage_oob: !MSG_PEEK skb: %p, users: %d, len: %d, consumed: %d\n", current->pid, (void*)skb, refcount_read(&skb->users), skb->len, UNIXCB(skb).consumed);
 				skb_unlink(skb, &sk->sk_receive_queue);
 				consume_skb(skb);
 				skb = skb_peek(&sk->sk_receive_queue);
@@ -2745,6 +2763,7 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 
 redo:
 		unix_state_lock(sk);
+		// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: get lock: %p", current->pid, (void*)sk);
 		if (sock_flag(sk, SOCK_DEAD)) {
 			err = -ECONNRESET;
 			goto unlock;
@@ -2754,8 +2773,10 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 
 #if IS_ENABLED(CONFIG_AF_UNIX_OOB)
 		if (skb) {
+			// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: manage_oob skb: %p, users: %d, lock: %p", current->pid, (void*)skb, refcount_read(&skb->users), (void*)sk);
 			skb = manage_oob(skb, sk, flags, copied);
 			if (!skb) {
+				// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: release lock1: %p", current->pid, (void*)sk);
 				unix_state_unlock(sk);
 				if (copied)
 					break;
@@ -2777,7 +2798,6 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 				goto unlock;
 			if (sk->sk_shutdown & RCV_SHUTDOWN)
 				goto unlock;
-
 			unix_state_unlock(sk);
 			if (!timeo) {
 				err = -EAGAIN;
@@ -2810,7 +2830,7 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 			if (!skb)
 				goto again;
 		}
-
+		// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: release lock4: %p", current->pid, (void*)sk);
 		unix_state_unlock(sk);
 
 		if (check_creds) {
@@ -2873,6 +2893,7 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 				break;
 
 			skb_unlink(skb, &sk->sk_receive_queue);
+			// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: consume_skb skb: %p, users: %d\n", current->pid, (void*)skb, refcount_read(&skb->users));
 			consume_skb(skb);
 
 			if (scm.fp)
@@ -2895,6 +2916,7 @@ static int unix_stream_read_generic(struct unix_stream_read_state *state,
 			skb = skb_peek_next(skb, &sk->sk_receive_queue);
 			if (skb)
 				goto again;
+			// printk(KERN_INFO "[PID: %d]  unix_stream_read_generic: release lock5: %p", current->pid, (void*)sk);
 			unix_state_unlock(sk);
 			break;
 		}
diff --git a/syzkaller/asm/a.out.h b/syzkaller/asm/a.out.h
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/syzkaller/asm/mce.h b/syzkaller/asm/mce.h
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/syzkaller/asm/msr.h b/syzkaller/asm/msr.h
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/syzkaller/asm/prctl.h b/syzkaller/asm/prctl.h
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/syzkaller/stdarg.h b/syzkaller/stdarg.h
new file mode 100644
index 000000000000..bedd2a4be7f2
--- /dev/null
+++ b/syzkaller/stdarg.h
@@ -0,0 +1,8 @@
+
+#pragma once
+#define va_list __builtin_va_list
+#define va_start __builtin_va_start
+#define va_end __builtin_va_end
+#define va_arg __builtin_va_arg
+#define va_copy __builtin_va_copy
+#define __va_copy __builtin_va_copy
diff --git a/syzkaller/uapi/asm/msr.h b/syzkaller/uapi/asm/msr.h
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tools/objtool/libsubcmd/include/subcmd/exec-cmd.h b/tools/objtool/libsubcmd/include/subcmd/exec-cmd.h
new file mode 100644
index 000000000000..aba591b8d254
--- /dev/null
+++ b/tools/objtool/libsubcmd/include/subcmd/exec-cmd.h
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __SUBCMD_EXEC_CMD_H
+#define __SUBCMD_EXEC_CMD_H
+
+extern void exec_cmd_init(const char *exec_name, const char *prefix,
+			  const char *exec_path, const char *exec_path_env);
+
+extern void set_argv_exec_path(const char *exec_path);
+extern const char *extract_argv0_path(const char *path);
+extern void setup_path(void);
+extern int execv_cmd(const char **argv); /* NULL terminated */
+extern int execl_cmd(const char *cmd, ...);
+/* get_argv_exec_path and system_path return malloc'd string, caller must free it */
+extern char *get_argv_exec_path(void);
+extern char *system_path(const char *path);
+
+#endif /* __SUBCMD_EXEC_CMD_H */
diff --git a/tools/objtool/libsubcmd/include/subcmd/help.h b/tools/objtool/libsubcmd/include/subcmd/help.h
new file mode 100644
index 000000000000..355c066c8d49
--- /dev/null
+++ b/tools/objtool/libsubcmd/include/subcmd/help.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __SUBCMD_HELP_H
+#define __SUBCMD_HELP_H
+
+#include <sys/types.h>
+#include <stdio.h>
+
+struct cmdnames {
+	size_t alloc;
+	size_t cnt;
+	struct cmdname {
+		size_t len; /* also used for similarity index in help.c */
+		char name[];
+	} **names;
+};
+
+static inline void mput_char(char c, unsigned int num)
+{
+	while(num--)
+		putchar(c);
+}
+
+void load_command_list(const char *prefix,
+		struct cmdnames *main_cmds,
+		struct cmdnames *other_cmds);
+void add_cmdname(struct cmdnames *cmds, const char *name, size_t len);
+void clean_cmdnames(struct cmdnames *cmds);
+int cmdname_compare(const void *a, const void *b);
+void uniq(struct cmdnames *cmds);
+/* Here we require that excludes is a sorted list. */
+void exclude_cmds(struct cmdnames *cmds, struct cmdnames *excludes);
+int is_in_cmdlist(struct cmdnames *c, const char *s);
+void list_commands(const char *title, struct cmdnames *main_cmds,
+		   struct cmdnames *other_cmds);
+
+#endif /* __SUBCMD_HELP_H */
diff --git a/tools/objtool/libsubcmd/include/subcmd/pager.h b/tools/objtool/libsubcmd/include/subcmd/pager.h
new file mode 100644
index 000000000000..a818964693ab
--- /dev/null
+++ b/tools/objtool/libsubcmd/include/subcmd/pager.h
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __SUBCMD_PAGER_H
+#define __SUBCMD_PAGER_H
+
+extern void pager_init(const char *pager_env);
+
+extern void setup_pager(void);
+extern int pager_in_use(void);
+extern int pager_get_columns(void);
+extern void force_pager(const char *);
+
+#endif /* __SUBCMD_PAGER_H */
diff --git a/tools/objtool/libsubcmd/include/subcmd/parse-options.h b/tools/objtool/libsubcmd/include/subcmd/parse-options.h
new file mode 100644
index 000000000000..41b9b942504d
--- /dev/null
+++ b/tools/objtool/libsubcmd/include/subcmd/parse-options.h
@@ -0,0 +1,247 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __SUBCMD_PARSE_OPTIONS_H
+#define __SUBCMD_PARSE_OPTIONS_H
+
+#include <linux/kernel.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#ifndef NORETURN
+#define NORETURN __attribute__((__noreturn__))
+#endif
+
+enum parse_opt_type {
+	/* special types */
+	OPTION_END,
+	OPTION_ARGUMENT,
+	OPTION_GROUP,
+	/* options with no arguments */
+	OPTION_BIT,
+	OPTION_BOOLEAN,
+	OPTION_INCR,
+	OPTION_SET_UINT,
+	OPTION_SET_PTR,
+	/* options with arguments (usually) */
+	OPTION_STRING,
+	OPTION_INTEGER,
+	OPTION_LONG,
+	OPTION_ULONG,
+	OPTION_CALLBACK,
+	OPTION_U64,
+	OPTION_UINTEGER,
+};
+
+enum parse_opt_flags {
+	PARSE_OPT_KEEP_DASHDASH = 1,
+	PARSE_OPT_STOP_AT_NON_OPTION = 2,
+	PARSE_OPT_KEEP_ARGV0 = 4,
+	PARSE_OPT_KEEP_UNKNOWN = 8,
+	PARSE_OPT_NO_INTERNAL_HELP = 16,
+};
+
+enum parse_opt_option_flags {
+	PARSE_OPT_OPTARG  = 1,
+	PARSE_OPT_NOARG   = 2,
+	PARSE_OPT_NONEG   = 4,
+	PARSE_OPT_HIDDEN  = 8,
+	PARSE_OPT_LASTARG_DEFAULT = 16,
+	PARSE_OPT_DISABLED = 32,
+	PARSE_OPT_EXCLUSIVE = 64,
+	PARSE_OPT_NOEMPTY  = 128,
+	PARSE_OPT_NOBUILD  = 256,
+	PARSE_OPT_CANSKIP  = 512,
+};
+
+struct option;
+typedef int parse_opt_cb(const struct option *, const char *arg, int unset);
+
+/*
+ * `type`::
+ *   holds the type of the option, you must have an OPTION_END last in your
+ *   array.
+ *
+ * `short_name`::
+ *   the character to use as a short option name, '\0' if none.
+ *
+ * `long_name`::
+ *   the long option name, without the leading dashes, NULL if none.
+ *
+ * `value`::
+ *   stores pointers to the values to be filled.
+ *
+ * `argh`::
+ *   token to explain the kind of argument this option wants. Keep it
+ *   homogeneous across the repository.
+ *
+ * `help`::
+ *   the short help associated to what the option does.
+ *   Must never be NULL (except for OPTION_END).
+ *   OPTION_GROUP uses this pointer to store the group header.
+ *
+ * `flags`::
+ *   mask of parse_opt_option_flags.
+ *   PARSE_OPT_OPTARG: says that the argument is optional (not for BOOLEANs)
+ *   PARSE_OPT_NOARG: says that this option takes no argument, for CALLBACKs
+ *   PARSE_OPT_NONEG: says that this option cannot be negated
+ *   PARSE_OPT_HIDDEN this option is skipped in the default usage, showed in
+ *                    the long one.
+ *
+ * `callback`::
+ *   pointer to the callback to use for OPTION_CALLBACK.
+ *
+ * `defval`::
+ *   default value to fill (*->value) with for PARSE_OPT_OPTARG.
+ *   OPTION_{BIT,SET_UINT,SET_PTR} store the {mask,integer,pointer} to put in
+ *   the value when met.
+ *   CALLBACKS can use it like they want.
+ *
+ * `set`::
+ *   whether an option was set by the user
+ */
+struct option {
+	enum parse_opt_type type;
+	int short_name;
+	const char *long_name;
+	void *value;
+	const char *argh;
+	const char *help;
+	const char *build_opt;
+
+	int flags;
+	parse_opt_cb *callback;
+	intptr_t defval;
+	bool *set;
+	void *data;
+	const struct option *parent;
+};
+
+#define check_vtype(v, type) ( BUILD_BUG_ON_ZERO(!__builtin_types_compatible_p(typeof(v), type)) + v )
+
+#define OPT_END()                   { .type = OPTION_END }
+#define OPT_PARENT(p)               { .type = OPTION_END, .parent = (p) }
+#define OPT_ARGUMENT(l, h)          { .type = OPTION_ARGUMENT, .long_name = (l), .help = (h) }
+#define OPT_GROUP(h)                { .type = OPTION_GROUP, .help = (h) }
+#define OPT_BIT(s, l, v, h, b)      { .type = OPTION_BIT, .short_name = (s), .long_name = (l), .value = check_vtype(v, int *), .help = (h), .defval = (b) }
+#define OPT_BOOLEAN(s, l, v, h)     { .type = OPTION_BOOLEAN, .short_name = (s), .long_name = (l), .value = check_vtype(v, bool *), .help = (h) }
+#define OPT_BOOLEAN_FLAG(s, l, v, h, f)     { .type = OPTION_BOOLEAN, .short_name = (s), .long_name = (l), .value = check_vtype(v, bool *), .help = (h), .flags = (f) }
+#define OPT_BOOLEAN_SET(s, l, v, os, h) \
+	{ .type = OPTION_BOOLEAN, .short_name = (s), .long_name = (l), \
+	.value = check_vtype(v, bool *), .help = (h), \
+	.set = check_vtype(os, bool *)}
+#define OPT_INCR(s, l, v, h)        { .type = OPTION_INCR, .short_name = (s), .long_name = (l), .value = check_vtype(v, int *), .help = (h) }
+#define OPT_SET_UINT(s, l, v, h, i)  { .type = OPTION_SET_UINT, .short_name = (s), .long_name = (l), .value = check_vtype(v, unsigned int *), .help = (h), .defval = (i) }
+#define OPT_SET_PTR(s, l, v, h, p)  { .type = OPTION_SET_PTR, .short_name = (s), .long_name = (l), .value = (v), .help = (h), .defval = (p) }
+#define OPT_INTEGER(s, l, v, h)     { .type = OPTION_INTEGER, .short_name = (s), .long_name = (l), .value = check_vtype(v, int *), .help = (h) }
+#define OPT_UINTEGER(s, l, v, h)    { .type = OPTION_UINTEGER, .short_name = (s), .long_name = (l), .value = check_vtype(v, unsigned int *), .help = (h) }
+#define OPT_UINTEGER_OPTARG(s, l, v, d, h)    { .type = OPTION_UINTEGER, .short_name = (s), .long_name = (l), .value = check_vtype(v, unsigned int *), .help = (h), .flags = PARSE_OPT_OPTARG, .defval = (intptr_t)(d) }
+#define OPT_LONG(s, l, v, h)        { .type = OPTION_LONG, .short_name = (s), .long_name = (l), .value = check_vtype(v, long *), .help = (h) }
+#define OPT_ULONG(s, l, v, h)        { .type = OPTION_ULONG, .short_name = (s), .long_name = (l), .value = check_vtype(v, unsigned long *), .help = (h) }
+#define OPT_U64(s, l, v, h)         { .type = OPTION_U64, .short_name = (s), .long_name = (l), .value = check_vtype(v, u64 *), .help = (h) }
+#define OPT_STRING(s, l, v, a, h)   { .type = OPTION_STRING,  .short_name = (s), .long_name = (l), .value = check_vtype(v, const char **), .argh = (a), .help = (h) }
+#define OPT_STRING_OPTARG(s, l, v, a, h, d) \
+	{ .type = OPTION_STRING,  .short_name = (s), .long_name = (l), \
+	  .value = check_vtype(v, const char **), .argh =(a), .help = (h), \
+	  .flags = PARSE_OPT_OPTARG, .defval = (intptr_t)(d) }
+#define OPT_STRING_OPTARG_SET(s, l, v, os, a, h, d) \
+	{ .type = OPTION_STRING, .short_name = (s), .long_name = (l), \
+	  .value = check_vtype(v, const char **), .argh = (a), .help = (h), \
+	  .flags = PARSE_OPT_OPTARG, .defval = (intptr_t)(d), \
+	  .set = check_vtype(os, bool *)}
+#define OPT_STRING_NOEMPTY(s, l, v, a, h)   { .type = OPTION_STRING,  .short_name = (s), .long_name = (l), .value = check_vtype(v, const char **), .argh = (a), .help = (h), .flags = PARSE_OPT_NOEMPTY}
+#define OPT_DATE(s, l, v, h) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), .value = (v), .argh = "time", .help = (h), .callback = parse_opt_approxidate_cb }
+#define OPT_CALLBACK(s, l, v, a, h, f) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), .value = (v), .argh = (a), .help = (h), .callback = (f) }
+#define OPT_CALLBACK_SET(s, l, v, os, a, h, f) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), .value = (v), .argh = (a), .help = (h), .callback = (f), .set = check_vtype(os, bool *)}
+#define OPT_CALLBACK_NOOPT(s, l, v, a, h, f) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), .value = (v), .argh = (a), .help = (h), .callback = (f), .flags = PARSE_OPT_NOARG }
+#define OPT_CALLBACK_DEFAULT(s, l, v, a, h, f, d) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), .value = (v), .argh = (a), .help = (h), .callback = (f), .defval = (intptr_t)d, .flags = PARSE_OPT_LASTARG_DEFAULT }
+#define OPT_CALLBACK_DEFAULT_NOOPT(s, l, v, a, h, f, d) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l),\
+	.value = (v), .arg = (a), .help = (h), .callback = (f), .defval = (intptr_t)d,\
+	.flags = PARSE_OPT_LASTARG_DEFAULT | PARSE_OPT_NOARG}
+#define OPT_CALLBACK_OPTARG(s, l, v, d, a, h, f) \
+	{ .type = OPTION_CALLBACK, .short_name = (s), .long_name = (l), \
+	  .value = (v), .argh = (a), .help = (h), .callback = (f), \
+	  .flags = PARSE_OPT_OPTARG, .data = (d) }
+
+/* parse_options() will filter out the processed options and leave the
+ * non-option argments in argv[].
+ * Returns the number of arguments left in argv[].
+ *
+ * NOTE: parse_options() and parse_options_subcommand() may call exit() in the
+ * case of an error (or for 'special' options like --list-cmds or --list-opts).
+ */
+extern int parse_options(int argc, const char **argv,
+                         const struct option *options,
+                         const char * const usagestr[], int flags);
+
+extern int parse_options_subcommand(int argc, const char **argv,
+				const struct option *options,
+				const char *const subcommands[],
+				const char *usagestr[], int flags);
+
+extern NORETURN void usage_with_options(const char * const *usagestr,
+                                        const struct option *options);
+extern NORETURN __attribute__((format(printf,3,4)))
+void usage_with_options_msg(const char * const *usagestr,
+			    const struct option *options,
+			    const char *fmt, ...);
+
+/*----- incremantal advanced APIs -----*/
+
+enum {
+	PARSE_OPT_HELP = -1,
+	PARSE_OPT_DONE,
+	PARSE_OPT_LIST_OPTS,
+	PARSE_OPT_LIST_SUBCMDS,
+	PARSE_OPT_UNKNOWN,
+};
+
+/*
+ * It's okay for the caller to consume argv/argc in the usual way.
+ * Other fields of that structure are private to parse-options and should not
+ * be modified in any way.
+ */
+struct parse_opt_ctx_t {
+	const char **argv;
+	const char **out;
+	int argc, cpidx;
+	const char *opt;
+	const struct option *excl_opt;
+	int flags;
+};
+
+extern int parse_options_usage(const char * const *usagestr,
+			       const struct option *opts,
+			       const char *optstr,
+			       bool short_opt);
+
+
+/*----- some often used options -----*/
+extern int parse_opt_abbrev_cb(const struct option *, const char *, int);
+extern int parse_opt_approxidate_cb(const struct option *, const char *, int);
+extern int parse_opt_verbosity_cb(const struct option *, const char *, int);
+
+#define OPT__VERBOSE(var)  OPT_BOOLEAN('v', "verbose", (var), "be verbose")
+#define OPT__QUIET(var)    OPT_BOOLEAN('q', "quiet",   (var), "be quiet")
+#define OPT__VERBOSITY(var) \
+	{ OPTION_CALLBACK, 'v', "verbose", (var), NULL, "be more verbose", \
+	  PARSE_OPT_NOARG, &parse_opt_verbosity_cb, 0 }, \
+	{ OPTION_CALLBACK, 'q', "quiet", (var), NULL, "be more quiet", \
+	  PARSE_OPT_NOARG, &parse_opt_verbosity_cb, 0 }
+#define OPT__DRY_RUN(var)  OPT_BOOLEAN('n', "dry-run", (var), "dry run")
+#define OPT__ABBREV(var)  \
+	{ OPTION_CALLBACK, 0, "abbrev", (var), "n", \
+	  "use <n> digits to display SHA-1s", \
+	  PARSE_OPT_OPTARG, &parse_opt_abbrev_cb, 0 }
+
+extern const char *parse_options_fix_filename(const char *prefix, const char *file);
+
+void set_option_flag(struct option *opts, int sopt, const char *lopt, int flag);
+void set_option_nobuild(struct option *opts, int shortopt, const char *longopt,
+			const char *build_opt, bool can_skip);
+
+#endif /* __SUBCMD_PARSE_OPTIONS_H */
diff --git a/tools/objtool/libsubcmd/include/subcmd/run-command.h b/tools/objtool/libsubcmd/include/subcmd/run-command.h
new file mode 100644
index 000000000000..17d969c6add3
--- /dev/null
+++ b/tools/objtool/libsubcmd/include/subcmd/run-command.h
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __SUBCMD_RUN_COMMAND_H
+#define __SUBCMD_RUN_COMMAND_H
+
+#include <unistd.h>
+
+enum {
+	ERR_RUN_COMMAND_FORK = 10000,
+	ERR_RUN_COMMAND_EXEC,
+	ERR_RUN_COMMAND_PIPE,
+	ERR_RUN_COMMAND_WAITPID,
+	ERR_RUN_COMMAND_WAITPID_WRONG_PID,
+	ERR_RUN_COMMAND_WAITPID_SIGNAL,
+	ERR_RUN_COMMAND_WAITPID_NOEXIT,
+};
+#define IS_RUN_COMMAND_ERR(x) (-(x) >= ERR_RUN_COMMAND_FORK)
+
+struct child_process {
+	const char **argv;
+	pid_t pid;
+	/*
+	 * Using .in, .out, .err:
+	 * - Specify 0 for no redirections (child inherits stdin, stdout,
+	 *   stderr from parent).
+	 * - Specify -1 to have a pipe allocated as follows:
+	 *     .in: returns the writable pipe end; parent writes to it,
+	 *          the readable pipe end becomes child's stdin
+	 *     .out, .err: returns the readable pipe end; parent reads from
+	 *          it, the writable pipe end becomes child's stdout/stderr
+	 *   The caller of start_command() must close the returned FDs
+	 *   after it has completed reading from/writing to it!
+	 * - Specify > 0 to set a channel to a particular FD as follows:
+	 *     .in: a readable FD, becomes child's stdin
+	 *     .out: a writable FD, becomes child's stdout/stderr
+	 *     .err > 0 not supported
+	 *   The specified FD is closed by start_command(), even in case
+	 *   of errors!
+	 */
+	int in;
+	int out;
+	int err;
+	const char *dir;
+	const char *const *env;
+	unsigned no_stdin:1;
+	unsigned no_stdout:1;
+	unsigned no_stderr:1;
+	unsigned exec_cmd:1; /* if this is to be external sub-command */
+	unsigned stdout_to_stderr:1;
+	void (*preexec_cb)(void);
+};
+
+int start_command(struct child_process *);
+int finish_command(struct child_process *);
+int run_command(struct child_process *);
+
+#define RUN_COMMAND_NO_STDIN 1
+#define RUN_EXEC_CMD	     2	/*If this is to be external sub-command */
+#define RUN_COMMAND_STDOUT_TO_STDERR 4
+int run_command_v_opt(const char **argv, int opt);
+
+#endif /* __SUBCMD_RUN_COMMAND_H */
